### DISCUSSION TITLE
Salt orchestrator breaks when master is restarted, even with high availability

### DISCUSSION THREAD

**Discussion by deepak-goel** [saltstack/salt#67410] [Feature Requests]
### Description of Issue
If we're deploying saltstack code in docker container using a multi-master approach, everything works fine wrt high availability if minions run a `salt-call` and one of the masters go down (since there is a retry mechanism on the minion side)

But if we've started an orchestrator from one of the docker containers, and mid-way due to a saltstack code deployment, the container is killed and a new one spawns up with the new code, we lose all the progress AND there is no way for high availability for orchestrator. After reading up the documentation, I agree that we can store the minion/job cache in an external database (mysql/consul etc), or use a returner for this, but there's no process which resumes/picks-up the orchestrator incase one of the masters go down

PS - 
1. salt-syndic also does not solve this approach since if one of the master-of-masters (MoM) is restarted during an orchestration-run, we end up with the same issue.
2. the issue isn't obviously specific to docker. We end up with the same issue when there's a traditional saltstack setup, and the `salt-master` process is restarted after every new code deployment.

### Setup
NA

### Steps to Reproduce Issue
* Have a multi-master saltstack setup
* Start an orchestrator from one of the masters
* Restart the same master before the orchestrator is finished
* The orchestrator is killed midway without a possibility for delegation to the other salt-master (which is expected 'coz the setup is supposed to be HA)

### Versions Report
NA

### Suggestion
Since there's already a process to store the minion/job cache on an external system, we could have a watcher process on the salt-masters which keep tracks of such use-cases and if one of the masters go down, the other master is automatically delegated that job.

cc @yagnik

**Comment by deepak-goel** [comment]
@Akm0d is this something that we can work on?

**Comment by whytewolf** [comment]
the ask here is a lot more work than benefit. and the simplistic solution provided by a watcher is problematic in most cases since masters in HA know nothing of each other. this watcher would need to be more than just a simple watching process it would also need to be a bridge between masters. 

Something more like a celery task queue might be better and fixes a few other problems. As well as removes the need to add code that has masters know about other masters. However, a celery queue adds the risk of security to the problem. as that becomes a way of adding a task into the masters. so how do you limit how tasks are added to the queue. 

However, that doesn't fix the overall problem that the original question should bring up but doesn't. What happens to processes that are in the middle of execution when a master is restarted? for some they will want to rerun that process, for others they may not want to, or have to run another orchestration to get everything into a known state before running further. still, others might need to stop the entire orchestration. scrap everything that was done and start over. I'm sure there are other possibilities that I am not thinking about. and what's worse is that all possible cases might exist within the same setups. these would require adding information into the orchestration to handle these situations. which adds complexity, 

The issue arises that this can already be taken care of with proper monitoring triggering events and orchestration that is idempotent. as well as not over-designing orchestrations. if it takes more than an hour for your orchestration to run. it might be better to break it up into reactive bits. say updating a Cassandra node. that should be three independent steps. an orchestration that removes it from the cluster. a state that updates it and then another orchestration that adds that node back in. the first orchestration can fire an event that starts the state that updates it. and then the state can fire an event when needed that causes the node readd orchestration.

**Comment by deepak-goel** [comment]
@whytewolf While I agree to some bits of your comment, I still think there's a gap when we look at HA for salt-orchestration.

Whether we need to use celery or something else is a thing for later, but acknowledging the fact that salt-orchestration needs to have better availability is what's more important for the community. I'm pretty sure there are lots of people who've long salt orchestrations, and as a workaround they would actually be following your solution of breaking one big orchestration into multiple small ones. While that addresses the current problem, but this solution makes it more complex than it should be.

Can the core team please think about whether this is something they'd like to address? Maybe ask this to the community on slack or during our public sessions? I think making orchestrator truly HA will make saltstack even more dependable as a solution as compared to other tools.
